Fix reconstruction of CSI volumes

jsafrane opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
During CSI volume reconstruction it's not possible to tell, if the volume is attachable or not - CSIDriver instance may not be available, because kubelet may not have connection to the API server at that time.
 
Therefore remove all calls to `FindAttachablePluginBySpec` here:

* Reorder `reconstructVolume` not to call `FindAttachablePluginBySpec` for CSI volumes. It's just a simple refactoring.
* Add *uncertain attach-abilty* state during reconstruction, so volume reconstruction can finish without an API server.  When the API server becomes available, then set attach-ability based on CSIDriver instances and allow unmounting of volumes.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes: #117745

#### Special notes for your reviewer:

~WIP:~
* ~PR for discussion~
* ~missing unit tests~ Unit tests added
* E2e test added in https://github.com/kubernetes/kubernetes/pull/118097

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed kubelet startup getting stuck with `NewVolumeManagerReconstruction` feature enabled and a CSI volume present in /var/lib/kubelet/pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jsafrane committed:
Refactor FindAttachablePluginBySpec out of CSI code path

reconstructVolume() is called when kubelet may not have connection to the
API server yet, therefore it cannot get CSIDriver instances to figure out
if a CSI volume is attachable or not.

Refactor reconstructVolume(), so it does not need
FindAttachablePluginBySpec for CSI volumes, because all of them are
deviceMountable (i.e. FindDeviceMountablePluginBySpec always returns the
CSI volume plugin).

jsafrane committed:
Add uncertain state of volume attach-ability

During CSI volume reconstruction it's not possible to tell, if the volume
is attachable or not - CSIDriver instance may not be available, because
kubelet may not have connection to the API server at that time.

Adding uncertain state during reconstruction + adding a correct state when
the API server is available.

jsafrane committed:
Update volumesInUse after attachability is confirmed

node.status.volumesInUse should report only attachable volumes, therefore
it needs to wait for the reconciler to update uncertain attachability of
volumes from the API server.

jsafrane commented:
Added unit test and removed WIP:

bart0sh commented:
/triage accepted
/priority important-soon

jsafrane commented:
/test help

bart0sh commented:
/sig storage

jingxu97 commented:
@jsafrane would you mind adding some more details about the issue and fix?

jsafrane commented:
@jingxu97, there are details in the issue, https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534540981 + https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534837217

jsafrane commented:
And  I added some more details in the PR description. Check also individual commits - they should be relatively short.

jsafrane commented:
I added a new e2e job, `pull-kubernetes-e2e-storage-kind-disruptive`. It runs kube-apiserver as a static pod and tries to reproduce https://github.com/kubernetes/kubernetes/issues/117745#issuecomment-1534540981.

* With `NewVolumeManagerReconstruction` feature enabled without any change, it fails: https://github.com/kubernetes/kubernetes/pull/118410
* With this PR + `NewVolumeManagerReconstruction` feature enabled  it succeeds: https://github.com/kubernetes/kubernetes/pull/118506

So, the issue should be fixed by this PR. I also tested manually with kube-apiserver as a static pod.

I'm proposing to review + merge this PR (or even https://github.com/kubernetes/kubernetes/pull/118506), let it soak in CI for a while and ask people who hit https://github.com/kubernetes/kubernetes/issues/117745 to test with master + `NewVolumeManagerReconstruction` enabled.

jsafrane commented:
/retest
CPU quota flake

dims commented:
/assign @jingxu97

msau42 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -361,7 +375,20 @@ type mountedPod struct {
-	return asw.addVolume(volumeName, volumeSpec, devicePath)
+
+	pluginIsAttachable := volumeAttachabilityFalse
+	if attachablePlugin, err := asw.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec); err == nil && attachablePlugin != nil {
```
Is there any concern of marking the volume attachability as false if there's a temporary error?

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_common.go:
```diff
@@ -236,17 +236,25 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (rvolume *reconstructe
 	if err != nil {
 		return nil, err
 	}
+	needsNameFromSpec := deviceMountablePlugin != nil
```
Can we document why we need name from spec if volume plugin is attachable or mountable?

jsafrane replied:
I copied comment from DSWP: https://github.com/kubernetes/kubernetes/blob/1f2d49972cc541606b7a2a902271aa293b836d17/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L279-L283
The unique volume name must be reconstructed the same way.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct_common.go:
```diff
@@ -236,17 +236,25 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (rvolume *reconstructe
+			return nil, err
+		}
+		needsNameFromSpec = attachablePlugin != nil
+	}
```
Yes, this particular fix (the fist commit) just works around calling `FindAttachablePluginBySpec` for CSI volumes. CSI volumes are always device mountable, so they won't reach `FindAttachablePluginBySpec` here.

jsafrane commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -361,7 +375,20 @@ type mountedPod struct {
+}
+
+func (asw *actualStateOfWorld) AddReconstructedVolume(
+	logger klog.Logger,
```
Removed

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:
```diff
@@ -525,6 +525,54 @@ func TestActualStateOfWorld_FoundDuringReconstruction(t *testing.T) {
 			},
 		},
+		{
+			name: "uncertain attachability is resolved to attachable",
```
So when we first build ASOW during reconsturction without API server connectivity we can't know atttachability of CSI volumes and hence "attachable" property is uncertain.

So reconciler continuously tries to fix that state and assuming once kubelet is able to talk to api-server uncertain attachability will be resolved.

jsafrane commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -169,14 +169,20 @@ type ActualStateOfWorld interface {
 	SyncReconstructedVolume(volumeName v1.UniqueVolumeName, podName volumetypes.UniquePodName, outerVolumeSpecName string)
 
+	// Add the specified volume to ASW as uncertainly attached.
+	AddReconstructedVolume(volumeName v1.UniqueVolumeName, volumeSpec *volume.Spec, nodeName types.NodeName, devicePath string) error
```
Renamed to `AddAttachUncertainReconstructedVolume`

jsafrane committed:
Rename volumesNeedDevicePath

To volumesNeedUpdateFromNodeStatus - because both devicePath and uncertain
attach-ability needs to be fixed from node status.

jsafrane committed:
Rename updateReconstructedFromAPIServer

to be in sync with volumesNeedUpdateFromNodeStatus.

jsafrane commented:
I also renamed `volumesNeedDevicePath` to `volumesNeedUpdateFromNodeStatus` and `updateReconstructedFromAPIServer()` to `updateReconstructedFromNodeStatus` to reflect better what they actually do - see the last 2 commits.

jsafrane commented:
/retest

jingxu97 commented:
/lgtm
